### PR TITLE
No governance after cage

### PR DESCRIPTION
**Pull request by gbalabasquer**
This is a very early draft of how could look a `pool` that excludes governance after general shutdown.
This actual code doesn't work with `quit`, as `nope`s the old `hub` in the same `file` function. However we are also evaluating to remove `quit` in favor of the spell doing 2 `grab`s.

### CODE REVIEW

**Comment by gbalabasquer** [comment]
I think the PR is ready to be reviewed.

**Comment by gbalabasquer** [comment]
Tests moved, this should be fully ready for review.

File: src/tests/stubs/D3MTestPool.sol [hunk 1]
```diff
@@ -62,6 +72,11 @@ contract D3MTestPool is ID3MPool {
         _;
     }
 
+    modifier onlyHub {
+        require(msg.sender == hub, "D3MAavePool/only-hub");
```
Comment (iamchrissmith) [inline src/tests/stubs/D3MTestPool.sol:76]:
```suggestion
        require(msg.sender == hub, "D3MTestPool/only-hub");
```

**Review by iamchrissmith — COMMENTED** [review]
one minor change.  i want to look at the files outside the PR diff, but i think this is looking good

**Review by iamchrissmith — APPROVED** [review]
i think this all looks good

File: src/pools/D3MAavePool.sol [hunk 2]
```diff
@@ -126,27 +137,26 @@ contract D3MAavePool is ID3MPool {
         emit Rely(usr);
     }
     function deny(address usr) external auth {
+        require(vat.live() == 1, "D3MAavePool/no-deny-during-shutdown");
```
Comment (talbaneth) [inline src/pools/D3MAavePool.sol:140]: why is this needed?

File: src/pools/D3MAavePool.sol [hunk 3]
```diff
@@ -126,27 +137,26 @@ contract D3MAavePool is ID3MPool {
...
 
     function file(bytes32 what, address data) external auth {
-        if (what == "king") king = data;
+        require(vat.live() == 1, "D3MAavePool/no-file-during-shutdown");
+        if (what == "hub") {
+            vat.nope(hub);
+            hub = data;
+            vat.hope(data);
+        }
+        else if (what == "king") king = data;
```
Comment (talbaneth) [inline src/pools/D3MAavePool.sol:152]:
```suggestion
        } else if (what == "king") king = data;
```

File: src/pools/ID3MPool.sol [hunk 4]
```diff
@@ -26,18 +26,6 @@ pragma solidity ^0.8.14;
     guidelines for assets/shares/maxWithdraw etc.
 */
 interface ID3MPool {
-    /**
```
Comment (talbaneth) [inline src/pools/ID3MPool.sol:29]: Consider changing the sentence `msg.sender must be authorized.` in `depost`, `withdraw` and `transfer` to `msg.sender must be the hub`.

File: src/pools/D3MAavePool.sol [hunk 5]
```diff
@@ -100,7 +103,10 @@ contract D3MAavePool is ID3MPool {
 
         TokenLike(dai_).approve(pool_, type(uint256).max);
 
-        CanLike(D3mHubLike(hub_).vat()).hope(hub_);
+        hub = hub_;
+
```
Comment (talbaneth) [inline src/pools/D3MAavePool.sol:107]:
```suggestion
```
nit - remove empty line.

File: src/pools/D3MAavePool.t.sol [hunk 6]
```diff
@@ -199,23 +204,36 @@ contract D3MAavePoolTest is D3MPoolBaseTest {
         assertEq(code, 0);
     }
 
+    function testFail_deposit_notHub() public {
+        D3MAavePool(d3mTestPool).deposit(1);
```
Comment (talbaneth) [inline src/pools/D3MAavePool.t.sol:208]:
```suggestion
        D3MTestGem(address(adai)).rely(address(aavePool));
        D3MAavePool(d3mTestPool).deposit(1);
```
Without this addition the test fails anyway.

**Review by talbaneth — COMMENTED** [review]
Finished going over the changes, please see the comments and a question.

File: src/pools/D3MAavePool.sol (hunk 2 above)
Comment (gbalabasquer) [reply src/pools/D3MAavePool.sol:140]: You are right, this is not needed as the `hub` uses the other modifier.

File: src/pools/D3MAavePool.sol [hunk 7]
```diff
@@ -100,7 +103,9 @@ contract D3MAavePool is ID3MPool {
 
         TokenLike(dai_).approve(pool_, type(uint256).max);
 
-        CanLike(D3mHubLike(hub_).vat()).hope(hub_);
+        hub = hub_;
+        VatLike vat_ = vat = VatLike(D3mHubLike(hub_).vat());
+        vat_.hope(hub_);
```
Comment (hexonaut) [inline src/pools/D3MAavePool.sol:108]: Don't need to store `vat_` on the stack after Solidity 0.8. Can just reference the immutable `vat` on line 108.

File: src/pools/D3MAavePool.sol (hunk 2 above)
Comment (gbalabasquer) [reply src/pools/D3MAavePool.sol]: Removed.

File: src/pools/D3MAavePool.sol (hunk 7 above)
Comment (gbalabasquer) [reply src/pools/D3MAavePool.sol]: Oh great, didn't know about it. Modified!

File: src/pools/D3MAavePool.t.sol (hunk 6 above)
Comment (gbalabasquer) [reply src/pools/D3MAavePool.t.sol:208]: This is a bit weird. It seems the last version doesn't have it but doesn't mark it as outdated. And I haven't changed it lately so not sure why you found it.

Comment (talbaneth) [reply src/pools/D3MAavePool.t.sol]: It was moved to D3MPoolBase.t.sol but I assume the same problem exists there.
It's not so optimal now as the fail tests in the base do not come together with the success tests.

Comment (gbalabasquer) [reply src/pools/D3MAavePool.t.sol]: Yeah that's a negative trade off. But at least we can enforce every pool will implement those reverts correctly.

**Review by iamchrissmith — APPROVED** [review]
changes look good

**Review by talbaneth — APPROVED** [review]
LGTM

**Review by godsflaw — APPROVED** [review]

File: src/pools/D3MAavePool.sol [hunk 8]
```diff
@@ -131,22 +141,19 @@ contract D3MAavePool is ID3MPool {
...
-    }
-
-    function nope(address hub) external override auth {
-        CanLike(D3mHubLike(hub).vat()).nope(hub);
-    }
-
     // Deposits Dai to Aave in exchange for adai which gets sent to the msg.sender
     // Aave: https://docs.aave.com/developers/v/2.0/the-core-protocol/lendingpool#deposit
-    function deposit(uint256 wad) external override auth {
+    function deposit(uint256 wad) external override onlyHub {
```
[the remark this answers is not in the document]
Comment (gbalabasquer) [reply src/pools/D3MAavePool.sol:156]: Yep!
